istioctl/pkg/auth/upgrade.go

```diff
@@ -36,31 +37,27 @@ import (
 	rbacproto "istio.io/api/rbac/v1alpha1"
 )
 
+// WorkloadLabels is the workload labels, for example, app: productpage.
+type WorkloadLabels map[string]string
+
 // ServiceToWorkloadLabels maps the short service name to the workload labels that it's pointing to.
 // This service is defined in same namespace as the ServiceRole that's using it.
 type ServiceToWorkloadLabels map[string]WorkloadLabels
 
-// WorkloadLabels is the workload labels, for example, app: productpage.
-type WorkloadLabels map[string]string
-
 type Upgrader struct {
-	IstioConfigStore model.ConfigStore
-	K8sClient        *kubernetes.Clientset
-	ServiceFiles     []string
-	// RoleNameToWorkloadLabels maps the ServiceRole name to the map of service name to the workload labels
-	// that the service pointing to.
-	// We could have remove the `service` key layer to make it ServiceRole name maps to workload labels.
-	// However, this is needed for unit tests.
-	RoleNameToWorkloadLabels map[string]ServiceToWorkloadLabels
-	V1PolicyFile             string
-	serviceRoles             []model.Config
-	serviceRoleBindings      []model.Config
+	K8sClient    *kubernetes.Clientset
+	ServiceFiles []string
+	// NamespaceToServiceToWorkloadLabels maps the namespace to service name to the workload labels
+	// that the service in this namespace pointing to.
+	NamespaceToServiceToWorkloadLabels map[string]ServiceToWorkloadLabels
+	V1PolicyFile                       string
+	serviceRoles                       []model.Config
+	serviceRoleBindings                []model.Config
+	ConvertedPolicies                  strings.Builder
```

Comment:
We should avoid to store the raw string in `Upgrader`, instead it's easier to separate the code into different components and add more support if we have the real `AuthorizationPolicy` Config in `Upgrader`, and create a separate function to do the serialization.
Could you put a TODO for refactor this?

Reply:
Could you explain more about the `separate the code into different components`? 

If you prefer the `Upgrader.UpgradeCRDs()` returns a string to the writer in `auth.go` like before, let me know.

Reply:
That means `Upgrader` will store the AuthorizationPolicy and converted ServiceRole as model.Config in it. There will be a separate function to output it to string when called.

Reply:
Doing this will require deep copying structs which can be expensive. For example, if we have a ServiceRole with 5 services in 1 rule, then for the referring ServiceRoleBinding, we need to create 5 AuthorizationPolicy with the same binding. I avoid deep copying by writing to `ConvertedPolicies` directly for each copy.

Another option to avoid deep copy and storing string in `Upgrader` is to output the string to the `auth` cmd, which was done before. Do you want to go with this?

Reply:
I don't think it will make any difference for user. Yes, it's a little more expensive, but it's just some auth policies which is usually less than 50 lines of text. How many policies do you process at a time? 10, 100 or 1000? Note this is called on command line and the time is not the key factor (why does it matter if it's 20ms slower?) What matters is a better code structure to make it easier for us to maintain and add more features in the long run.

Reply:
But why is it better in terms of code structure and maintenance to store converted ServiceRole and AuthorizationPolicy model.Config in `Upgrader`, rather than the converted policy in string format? In the end we just write it out to stdout, right?

Reply:
Generating config and serializing to string is 2 different task, it's separated everywhere if you take a look at the codebase. In the end we only need it in stdout but this is not the reason to put them all together.
We may want to do something else after it's generated, for example we can associate additional comment information when doing the serialization, we can output some summary information about the conversion. If they're stored as string you lose the semantic meaning too early.

One more thing, you don't really need to do deep copy at all, what's the issue of just storing the pointer of the config?

Reply:
As for deep copy, the reason is in each binding, they will refer to different ServiceRole, right? If we don't deep copy each binding, then `role` field in all bindings (in AuthorizationPolicy config) will refer to the latest ServiceRole.

Reply:
I mean as long as the config is generated in the memory, we can simply keep the pointer of it in `upgrader`, not?
Anyway, let's move forward on this PR, we can do this later.

Reply:
Thanks. Let's discuss this offline.